### ISSUE TITLE
New tags will be required for texlive 2016 (was: Texlive 2015 texlive.tlpdb not parsed by texscythe)

### ISSUE THREAD

**Issue by fishforbreakfast** [vext01/texscythe#2]
Hi,
## Thank you for this project.  I was using texscythe to create some tar archives of TeXLive 2015 which was just released.  Seems that there have been some additions to the tlpdb format as an error is given complaining about functions not being declared.  Here is my output:

  TeXScythe Version 0.1.1
##   (c) Edd Barrett 2013 vext01@gmail.com edd@openbsd.org

  sqldb: texscythe.db
  plist: PLIST
  prefix_filenames: 
  tlpdb: texlive.tlpdb
  arch: None
  dirs: True
  regex: None
  inc_pkgspecs: []
  exc_pkgspecs: []
  quiet: False
  skip_missing_archpkgs: False

Parsing TLPDB...Traceback (most recent call last):
  File "/usr/bin/texscyther", line 118, in <module>
    tlpdbparser.initdb(cfg)
  File "/usr/lib64/python2.7/site-packages/texscythe/tlpdbparser.py", line 238, in initdb
    parse(sess, cfg)
  File "/usr/lib64/python2.7/site-packages/texscythe/tlpdbparser.py", line 57, in parse
    parse_lines(sess, cfg, fh)
  File "/usr/lib64/python2.7/site-packages/texscythe/tlpdbparser.py", line 66, in parse_lines
    parse_line(sess, cfg, line.rstrip("\n"), state)
  File "/usr/lib64/python2.7/site-packages/texscythe/tlpdbparser.py", line 89, in parse_line
    (field, funcname))
texscythe.tlpdbparser.TeXParseError: Unknown package metadata 'relocated'. Missing handler parse_relocated_data()

I am not very familiar with python but I did manage to get rid of the error messages by adding some lines to the tlpdbparser.py file:

def parse_relocated_data(sess, cfg, data, state):
    pass

def parse_containersize_data(sess, cfg, data, state):
    pass

def parse_containermd5_data(sess, cfg, data, state):
    pass

def parse_doccontainersize_data(sess, cfg, data, state):
    pass

def parse_doccontainermd5_data(sess, cfg, data, state):
    pass

def parse_srccontainersize_data(sess, cfg, data, state):
    pass

def parse_srccontainermd5_data(sess, cfg, data, state):
    pass

def parse_catalogue_also_data(sess, cfg, data, state):
    pass

def parse_catalogue_topics_data(sess, cfg, data, state):
    pass

Then there were no more exceptions from the python interpreter but building the tar files failed (what should I expect?? silly me).  Anyway, Hope this helps.

Cheers,

Fred

**Comment by vext01** [comment]
Hey,

Thanks for the report.

I've not looked at texlive 2015 atall yet, but I will surely fix this when I do.

The failure you see is defensive programming (although the error message could be better). Basically I wanted texscythe to stop if it hit a directive it doesn't know what to do with. If these new directives in tl2015 do no affect the packing lists, then `pass` implementations are fine. But you are saying there is another crash?

**Comment by fishforbreakfast** [comment]
Edd,

Thank you for your reply.  I ran the shell script that calls texscythe 
with the modified texscythe.  Here is the output:

------------------------------------------------------------------------
   TeXScythe Version 0.1.1
   (c) Edd Barrett 2013 <vext01@gmail.com> <edd@openbsd.org>
------------------------------------------------------------------------
   sqldb: texscythe.db
   plist: PLIST
   prefix_filenames:
   tlpdb: texlive.tlpdb
   arch: None
   dirs: True
   regex: None
   inc_pkgspecs: []
   exc_pkgspecs: []
   quiet: False
   skip_missing_archpkgs: False

Parsing 
TLPDB...250...500...750...1000...1250...1500...1750...2000...2250...2500...2750...3000...3250...3500...3750...4000...4250...4500...4750...5000...5250...5500...5750...6000...

=========================
Database summary
=========================
Packages:         6020
Files:          147814
Dependencies:     3629
------------------------------------------------------------------------
   TeXScythe Version 0.1.1
   (c) Edd Barrett 2013 <vext01@gmail.com> <edd@openbsd.org>
------------------------------------------------------------------------
   sqldb: texscythe.db
   plist: tmplists/full
   prefix_filenames:
   tlpdb: texlive.tlpdb
   arch: None
   dirs: False
   regex: None
   inc_pkgspecs: ['scheme-full']
   exc_pkgspecs: ['scheme-full:doc', 'scheme-full:src']
   quiet: False
   skip_missing_archpkgs: False

Collecting include files:
   Building file list: done: scheme-full:run,src,doc,bin has 131924 files
Collecting exclude files:
   Building file list: done: scheme-full:doc has 32875 files
   Building file list: done: scheme-full:src has 5304 files
Performing subtract... Done
Sorting... Done
Writing file list to 'tmplists/full'...Done
------------------------------------------------------------------------
   TeXScythe Version 0.1.1
   (c) Edd Barrett 2013 <vext01@gmail.com> <edd@openbsd.org>
------------------------------------------------------------------------
   sqldb: texscythe.db
   plist: tmplists/docs
   prefix_filenames:
   tlpdb: texlive.tlpdb
   arch: None
   dirs: False
   regex: None
   inc_pkgspecs: ['scheme-full:doc']
   exc_pkgspecs: ['scheme-full:src', 'bibarts']
   quiet: False
   skip_missing_archpkgs: False

Collecting include files:
   Building file list: done: scheme-full:doc has 32875 files
Collecting exclude files:
   Building file list: done: scheme-full:src has 5304 files
   Building file list: done: bibarts:run,src,doc,bin has 8 files
Performing subtract... Done
Sorting... Done
Writing file list to 'tmplists/docs'...Done
------------------------------------------------------------------------
   TeXScythe Version 0.1.1
   (c) Edd Barrett 2013 <vext01@gmail.com> <edd@openbsd.org>
------------------------------------------------------------------------
   sqldb: texscythe.db
   plist: tmplists/src
   prefix_filenames:
   tlpdb: texlive.tlpdb
   arch: None
   dirs: False
   regex: None
   inc_pkgspecs: ['scheme-full:src']
   exc_pkgspecs: []
   quiet: False
   skip_missing_archpkgs: False

Collecting include files:
   Building file list: done: scheme-full:src has 5304 files
Collecting exclude files:
Performing subtract... Done
Sorting... Done
Writing file list to 'tmplists/src'...Done
Generating tarballs - please be patient...
         Creating tarballs/texlive-texmf-20150523.tar
tar: texmf-dist/scripts/pdfbook2/pdfbook2: Cannot stat: No such file or 
directory
tar: Exiting with failure status due to previous errors


Here is the call to texscythe:

texscyther \
   --nodirs \
   --subset \
     --include scheme-full \
     --exclude scheme-full:doc scheme-full:src \
   --output-plist tmplists/full

Hope this helps.

Cheers,

Fred

**Comment by vext01** [comment]
Where does the "Generating tarballs - please be patient..." part come from? That is your code?

What I think is happened (but I'm not sure) is that the texlive tlpdb references a file that does not exist. This has happened in the past. If this is the case, it's not a problem with texscythe, but the tlpdb.

**Comment by vext01** [comment]
Can you check the entry for that file in the tlpdb manually, then check to see if it exists in the texmf tree?

**Comment by fishforbreakfast** [comment]
Edd,

Found out that I downloaded the wrong texlive.tlpdb for the source version of texlive that I was working with.  I used the command:
svn co -r {20150523} svn://tug.org/texlive/trunk/Master/tlpkg
to check out the proper version of texlive.tlpdb.  Now texscythe works with the shell script that I have.  The tar error was from the script and not texscythe as you pointed out earlier.  Later versions of texlive.tlpdb do have the extra tags in them (06/24/2015, etc.).  Probably a development addition.

Probably, it is OK to close this issue.

Thank you again for your help!

Cheers,

Fred
